tensorboard/plugins/debugger/debugger_server_lib.py

```diff
@@ -0,0 +1,321 @@
+    try:
+      self._events_writer_manager.write_event(
+          tf.Event(
+              wall_time=0, step=0, file_version=constants.EVENTS_VERSION))
+    except IOError as e:
+      tf.logging.error(
+          "Writing to %s failed: %s",
+          self._events_writer_manager.get_current_file_name(), e)
+
+    # See if a backup file exists. If so, use it to initialize the registry.
+    self._registry_backup_file_path = os.path.join(
+        debugger_directory, constants.ALERT_REGISTRY_BACKUP_FILE_NAME)
+    initial_data = None
+
+    if tf.gfile.Exists(self._registry_backup_file_path):
+      # A backup file exists. Read its contents to use for initialization.
+      with tf.gfile.Open(self._registry_backup_file_path, "r") as backup_file:
+        backup_data_json_string = backup_file.read().strip()
+
+        try:
+          # Use the data to initialize the registry.
+          initial_data = json.loads(backup_data_json_string)
+        except ValueError as err:
+          # Could not parse the data. No backup data obtained.
+          tf.logging.error(
+              "Could not parse contents of %s: %s",
+              self._registry_backup_file_path, err)
+
+    self._numerics_alert_registry = numerics_alert.NumericsAlertRegistry(
+        initialization_list=initial_data)
+
+    self._numerics_alert_lock = threading.Lock()
+    curried_handler_constructor = functools.partial(
+        DebuggerDataStreamHandler,
+        self._events_writer_manager,
+        self._numerics_alert_callback)
+    grpc_debug_server.EventListenerBaseServicer.__init__(
+        self, receive_port, curried_handler_constructor)
+
+  def start_the_debugger_data_receiving_server(self):
+    """Starts the HTTP server for receiving health pills at `receive_port`.
+
+    After this method is called, health pills issued to host:receive_port
+    will be stored by this object. Calling this method also creates a file
+    within the log directory for storing health pill summary events.
+    """
+    self.run_server()
+
+  def get_events_file_name(self):
+    """Gets the name of the debugger events file currently being written to.
+
+    Returns:
+      The string name of the debugger events file currently being written to.
+      This is just the name of that file, not the full path to that file.
+    """
+    return self._events_writer_manager.get_current_file_name()
+
+  def _numerics_alert_callback(self, alert):
+    """Handles the case in which we receive a bad value (NaN, -/+ Inf).
+
+    Args:
+      alert: The alert to be registered.
+    """
+    self._numerics_alert_lock.acquire()
+    self._numerics_alert_registry.register(alert)
+    self._numerics_alert_lock.release()
+
+  def numerics_alert_report(self):
+    """Get a report of the numerics alerts that have occurred.
+
+    Returns:
+      A list of `numerics_alert.NumericsAlertReportRow`, sorted in ascending
+        order of first_timestamp.
+    """
+    self._numerics_alert_lock.acquire()
```

Comment:
why not `with self._numerics_alert_lock: return self._numerics_alert_lock_registry.report()`?